### PR TITLE
Handle DeletedFinalStateUnknown in NodeController

### LINKED ISSUE #34692: NodeController can't handle DeletedFinalStateUnknown it gets from controller framework.
cc @davidopp @wojtek-t @lavalamp @fgrzadkowski

### PR DESCRIPTION
**Pull request by gmarek**
Fix #34692

``` release-note
Fix panic in NodeController caused by receiving DeletedFinalStateUnknown object from the cache.
```

cc @davidopp

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34694)

### CODE REVIEW

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -294,7 +294,20 @@ func NewNodeController(
 					return
 				}
 
-				node := obj.(*api.Node)
+				node, isNode := obj.(*api.Node)
+				// We can get DeletedFinalStateUnknown instead of *api.Node here and we need to handle that correctly. #34692
+				if !isNode {
+					deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+					if !ok {
+						glog.Warningf("Received unexpected object: %v", obj)
```
Comment (wojtek-t) [inline pkg/controller/node/nodecontroller.go]: Errorf (if we don't get neither node nor deletedfinalstateunknown, then we really have problem)

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -294,7 +294,20 @@ func NewNodeController(
...
+						return
+					}
+					node, ok = deletedState.Obj.(*api.Node)
+					if !ok {
+						glog.Warningf("DeletedFinalStateUnknown contained non-Node object: %v", deletedState.Obj)
```
Comment (wojtek-t) [inline pkg/controller/node/nodecontroller.go]: Errorf

**Review by wojtek-t — COMMENTED** [review]
We should probably open an issue to go through the code and check we don't have the same problem in other places. @lavalamp - FYI

File: pkg/controller/node/nodecontroller.go (hunk 2 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Done.

**Comment by wojtek-t** [comment]
@gmarek - this PR probably requires release-note, as well as should probably be cherrrypicked

**Comment by gmarek** [comment]
Agreed

**Comment by davidopp** [comment]
Oops, sorry I missed this when I did #34786. Thanks for catching it.

**Pull request merged**